src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,533 @@ impl Display for SnapKey {
+                debug!("deleting {}", path);
+                if let Err(e) = fs::remove_file(&path) {
+                    warn!("failed to delete temporary file {}: {:?}", path, e);
+                }
             }
         }
     }
-}
 
-/// A reader that calculate checksum and verify it without read
-/// it from the beginning again.
-pub struct SnapValidationReader {
-    reader: File,
-    digest: Digest,
-    left: usize,
-    res: Option<u32>,
-}
+    /// A reader that calculate checksum and verify it without read
+    /// it from the beginning again.
+    pub struct SnapValidationReader {
+        reader: File,
+        digest: Digest,
+        left: usize,
+        res: Option<u32>,
+    }
+
+    impl SnapValidationReader {
+        /// Open the snap file located at specified path.
+        fn open<P: AsRef<Path>>(path: P) -> io::Result<SnapValidationReader> {
+            let reader = try!(File::open(path.as_ref()));
+            let digest = Digest::new(crc32::IEEE);
+            let len = try!(reader.metadata()).len();
+            if len < CRC32_BYTES_COUNT as u64 {
+                return Err(io::Error::new(ErrorKind::InvalidInput,
+                                          format!("file length {} < {}", len, CRC32_BYTES_COUNT)));
+            }
+            let left = len as usize - CRC32_BYTES_COUNT;
+            Ok(SnapValidationReader {
+                reader: reader,
+                digest: digest,
+                left: left,
+                res: None,
+            })
+        }
 
-impl SnapValidationReader {
-    /// Open the snap file located at specified path.
-    fn open<P: AsRef<Path>>(path: P) -> io::Result<SnapValidationReader> {
-        let reader = try!(File::open(path.as_ref()));
-        let digest = Digest::new(crc32::IEEE);
-        let len = try!(reader.metadata()).len();
-        if len < CRC32_BYTES_COUNT as u64 {
-            return Err(io::Error::new(ErrorKind::InvalidInput,
-                                      format!("file length {} < {}", len, CRC32_BYTES_COUNT)));
-        }
-        let left = len as usize - CRC32_BYTES_COUNT;
-        Ok(SnapValidationReader {
-            reader: reader,
-            digest: digest,
-            left: left,
-            res: None,
-        })
-    }
-
-    /// Validate the file
-    ///
-    /// If the reader will be consumed after calling this method, no further data can be
-    /// read from this reader again.
-    pub fn validate(&mut self) -> io::Result<()> {
-        if self.res.is_none() {
-            if self.left > 0 {
-                let cap = cmp::min(self.left, DEFAULT_READ_BUFFER_SIZE);
-                let mut buf = vec![0; cap];
-                while self.left > 0 {
-                    try!(self.read(&mut buf));
+        /// Validate the file
+        ///
+        /// If the reader will be consumed after calling this method, no further data can be
+        /// read from this reader again.
+        pub fn validate(&mut self) -> io::Result<()> {
```

Comment:
Grammatically, the reader is still on this scope, there is no way to prohibit any later calling.